src/reason_pprint_ast.ml

```diff
@@ -99,16 +99,61 @@ let exprDescrString x =
     (string_of_int x.pexp_loc.loc_end.Lexing.pos_lnum) ^
     "]"
 
+type ruleInfoData = {
+  reducePrecedence: precedence;
+  shiftPrecedence: precedence;
+}
+
+and ruleInfo =
+  (* Printing will be parsed with very high precedence, so not much need to
+     worry about ensuring it will reduce correctly. In short, you can put
+     `VeryHighPrecedence` content anywhere around an infix identifier without
+     wrapping in parens. For example `myFunc x y z` or `if x {y} else {z}`
+     The layout is kept in list form only to allow for elegant wrapping rules
+     to take into consideration the *number* of high precedence parsed items. *)
+  | VeryHighPrecedence of layoutNode list
+  (* Care should be taken to ensure the rule that caused it to be parsed will
+     reduce again on the printed output - context should carefully consider
+     wrapping in parens according to the ruleInfoData. *)
+  | SpecificPrecedence of ruleInfoData * layoutNode
+
+and associativity =
+  | Right
+  | Nonassoc
+  | Left
+
+and precedenceEntryType =
+  | TokenPrecedence
+  | CustomPrecedence
+
+and precedence =
+  | Token of string
+  | Custom of string
+
 (* Make a standard list *)
-type whenToDoSomething =
+and whenToDoSomething =
   | Never
   | IfNeed
   | Always
   (* Always_rec not only will break, it will break recursively up to the root *)
   | Always_rec
 
-type easyFormatLabelFormatter = Easy_format.t -> Easy_format.t -> Easy_format.t
-type listConfig = {
+(* Describes the "fixity" of a token, and stores its *printed* representation
+   should it be rendered as infix/prefix (This rendering may be different than
+   how it is stored in the AST). *)
+and tokenFixity =
+  (* Such as !simple_expr and ~!simple_expr. These function applications are
+     considered *almost* "simple" because they may be allowed anywhere a simple
+     expression is accepted, except for when on the left hand side of a
+     dot/send. *)
+  | AlmostSimplePrefix of string
```

Comment:
We should get rid of this, and have only unary precedence in the future. Once `!` is swapped with boolean inversion `not`, it will make more sense anyways that `! myFunc a b` should be parsed like unary prefix `!(myFunc a b)`, instead of like today's "almost simple prefix" `(!myFunc) a b`.